p2p/dandelion/src/txpool.rs

```diff
@@ -0,0 +1,545 @@
+//! # Dandelion++ TxPool
+//!
+//! This is an implementation of a dandelion++ compatible tx pool.
+//!
+use std::{
+    collections::{HashMap, HashSet},
+    future::Future,
+    hash::Hash,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{FutureExt, StreamExt};
+use rand::prelude::*;
+use rand_distr::Exp;
+use tokio::{
+    sync::{mpsc, oneshot},
+    task::JoinSet,
+};
+use tokio_util::{sync::PollSender, time::DelayQueue};
+use tower::{Service, ServiceExt};
+use tracing::Instrument;
+
+use crate::{
+    traits::{TxStoreRequest, TxStoreResponse},
+    DandelionConfig, DandelionRouteReq, DandelionRouterError, State, TxState,
+};
+
+#[derive(Copy, Clone, Debug, thiserror::Error)]
+#[error("The dandelion pool was shutdown")]
+pub struct DandelionPoolShutDown;
+
+/// A request to the dandelion tx pool.
+///
+/// This enum does not contain requests to get tx data from the pool, this is to reduce the amount
+/// of requests that get sent to the [`DandelionPool`]. Users should hold a handle to the backing tx storage
+/// to get transaction data from the pool.
+///
+/// It MUST be remembered to not expose the private stem pool to peers.
+///
+/// Users may notice there is no way to check if the dandelion-pool wants a tx according to an inventory message like seen
+/// in Bitcoin, only having a request for a full tx. Users should look in the *public* backing pool to handle inv messages,
+/// and request txs even if they are in the stem pool.
+pub enum DandelionPoolRequest<Tx, TxID, PID> {
+    /// An incoming tx.
+    IncomingTx {
+        /// The transaction.
+        tx: Tx,
+        /// The transaction ID.
+        tx_id: TxID,
+        /// The routing state of this transaction.
+        tx_state: TxState<PID>,
+    },
+}
```

Comment:
Will there be more variants in the future?

Reply:
no I'll make this a struct